ui/plugin-system/src/components/TimeSeriesQueryEditor.tsx

```diff
@@ -11,24 +11,137 @@
+import ChevronDown from 'mdi-material-ui/ChevronDown';
+import ChevronRight from 'mdi-material-ui/ChevronRight';
+import { TimeSeriesQueryDefinition, QueryDefinition } from '@perses-dev/core';
+import { usePlugin, usePluginRegistry } from '../runtime';
 import { PluginEditor, PluginEditorProps } from './PluginEditor';
 
+const DEFAULT_QUERY_PLUGIN_TYPE = 'TimeSeriesQuery';
+
+export interface TimeSeriesQueryEditorProps {
+  queries: TimeSeriesQueryDefinition[];
+  onChange: (queries: QueryDefinition[]) => void;
+}
+
+export function TimeSeriesQueryEditor(props: TimeSeriesQueryEditorProps) {
+  const { onChange, queries } = props;
+  const hasMoreThanOneQuery = queries.length > 1;
+  const { defaultPluginKinds } = usePluginRegistry();
+  const defaultTimeSeriesQueryKind = defaultPluginKinds?.[DEFAULT_QUERY_PLUGIN_TYPE] ?? '';
+
+  const { data: defaultQueryPlugin } = usePlugin(DEFAULT_QUERY_PLUGIN_TYPE, defaultTimeSeriesQueryKind, {
+    useErrorBoundary: true,
+    enabled: true,
+  });
+
+  // State for which queries are collapsed
+  // TODO: Would be easier if we had IDs for queries.
+  const [queriesCollapsed, setQueriesCollapsed] = useState(queries.map(() => false));
+
+  // Query handlers
+  const handleQueryChange = (index: number, queryDef: TimeSeriesQueryDefinition) => {
+    onChange(
+      produce(queries, (draft) => {
+        if (draft) {
+          draft[index] = queryDef;
+        } else {
+          draft = [queryDef];
+        }
+      })
+    );
+  };
+
+  const handleQueryAdd = () => {
+    if (!defaultQueryPlugin) return;
+    onChange(
+      produce(queries, (draft) => {
+        const queryDef: TimeSeriesQueryDefinition = {
+          kind: DEFAULT_QUERY_PLUGIN_TYPE,
+          spec: {
+            plugin: { kind: defaultTimeSeriesQueryKind, spec: defaultQueryPlugin.createInitialOptions() },
+          },
+        };
+        if (draft) {
+          draft.push(queryDef);
+        } else {
+          draft = [...queries, queryDef];
+        }
+      })
+    );
+    setQueriesCollapsed((queriesCollapsed) => {
+      queriesCollapsed.push(false);
+      return [...queriesCollapsed];
+    });
+  };
+
+  const handleQueryDelete = (index: number) => {
+    onChange(
+      produce(queries, (draft) => {
+        draft.splice(index, 1);
+      })
+    );
+    setQueriesCollapsed((queriesCollapsed) => {
+      queriesCollapsed.splice(index, 1);
+      return [...queriesCollapsed];
+    });
+  };
+
+  const handleQueryCollapseExpand = (index: number) => {
+    setQueriesCollapsed((queriesCollapsed) => {
+      queriesCollapsed[index] = !queriesCollapsed[index];
+      return [...queriesCollapsed];
+    });
+  };
+
+  return (
+    <Stack spacing={1}>
+      <Button variant="contained" startIcon={<AddIcon />} sx={{ marginLeft: 'auto' }} onClick={handleQueryAdd}>
+        Add Query
+      </Button>
+      {queries.map((query: TimeSeriesQueryDefinition, i: number) => (
+        <Stack key={i} spacing={1}>
```

Comment:
[optional suggestion] Would it make sense to turn the content of this loop into a reusable component? This file is pretty big & complex and that might help with maintenance/readability.

Reply:
converted this to `TimeSeriesQueryInput`